feat: add dev message for value sent to non-payable

unparalleled-js opened the pull request:
### What I did

adds custom dev message for payable sent to non-payable so it can be asserted on using `ape.reverts`.

### How I did it

copied brownie, sorta

### How to verify it

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

unparalleled-js commented on setup.py:
```diff
@@ -57,8 +57,8 @@
-        "eth-ape>=0.6.4,<0.7",
-        "ethpm-types>=0.4.1",  # Use same version as eth-ape
+        "eth-ape>=0.6.7,<0.7",
+        "ethpm-types>=0.4.2",  # Use same version as eth-ape
```
- [ ] use new, released `ethpm-types`: https://github.com/ApeWorX/ethpm-types/pull/69

unparalleled-js commented:
- [ ] not sure about something - should add more errors to be more sure

fubuloubu commented on ape_vyper/compiler.py:
```diff
@@ -20,6 +21,28 @@
 
 
+class DevMessage(Enum):
+    NONPAYABLE_CHECK = "Cannot send ether to non-payable function"
+    INDEX_OUT_OF_RANGE = "Index out of range"
+    INTEGER_OVERFLOW = "Integer overflow"
+    INTEGER_UNDERFLOW = "Integer underflow"
+    DIVISION_BY_ZERO = "Division by zero"
+    MODULO_BY_ZERO = "Modulo by zero"
```
Ah, okay you are doing these yourself? Can it be prepended with `dev: `?

unparalleled-js replied:
[044cfdf](https://github.com/ApeWorX/ape-vyper/pull/73/commits/044cfdf62253b6ec818abd037a4cc88736d0ef5f)

fubuloubu reviewed (commented):
Not to completely throw a monkey wrench in it, but I actually wonder if it might be better to raise a more descriptive error up the stack instead of raising as a dev assert message, like a built-in compiler plugin exception that it detects via pattern matching according to known compiler semantics

fubuloubu commented:
Otherwise, it's looking good to me

unparalleled-js commented:
> Not to completely throw a monkey wrench in it, but I actually wonder if it might be better to raise a more descriptive error up the stack instead of raising as a dev assert message, like a built-in compiler plugin exception that it detects via pattern matching according to known compiler semantics

Ya definitely would be ideal, just not sure exactly how to make that work. Also, would maybe break how ape.reverts works

fubuloubu commented:
> Also, would maybe break how ape.reverts works

Even if they subclasses `ContractLogicError`?

fubuloubu commented:
Just sketching an idea here, but it could look like taking the `ContractLogicError` exception, and then asking the compiler plugins if they can enrich that exception with a better error message that subclasses `ContractLogicError`

unparalleled-js commented:
@fubuloubu Ya, that may be pretty cool. Let's see what we can do...

I think I still will need the pc -> error map thing tho for the source trace back / coverage stuff.

unparalleled-js commented:
Love the `enrich_error` idea and have that in the works. Still need these error PCs somewhere tho, so I think for now, we should let both plans work.

Have others PRs now for enrich error but need this one to merge first.

unparalleled-js commented:
> Just sketching an idea here, but it could look like taking the `ContractLogicError` exception, and then asking the compiler plugins if they can enrich that exception with a better error message that subclasses `ContractLogicError`

Ok think I have it set up nicely here: https://github.com/ApeWorX/ape-vyper/pull/74
will be able to merge after next release.
In the mean time, I added a comment , dev prefix, and renamed the enum to try and find the best middle ground.
